File: adapters/ui-kotlin/src/main/kotlin/dev/flux/ui/ImageAdapter.kt
package dev.flux.ui

import java.lang.ref.WeakReference
import kotlin.collections.List as KList

/**
 * Declarative adapter for `Image` (unified tier; AGENTS.md §3.5).
 *
 * Maps a Flux `Image` node to a native image view, loading the bitmap from the
 * dev server's asset route (`http://localhost:7332/assets/<src>`). The adapter
 * is platform-neutral: it does not touch `android.*` directly, instead recording
 * the resolved request as [PROP_SRC] (and the optional `width`/`height`/
 * `contentMode` props) onto the [FluxNativeView]. The Android runtime (FLUX-007)
 * reads these properties and renders the actual `Image`/`painter`, handling the
 * HTTP fetch and graceful fallback to a placeholder (BR-003) on failure.
 *
 * Each node gets its own adapter instance via [create], so the resolved source
 * and dimensions never bleed into a sibling image (FLUX-007).
 *
 * Props are read by name; the index is the FNV-1a-32 digest of the name masked
 * to `u16` ([PropsIndex.propIndexForName]), derived identically on server and
 * client (AGENTS.md §3.2) — never a hardcoded positional index. Fields:
 * - `source: String` (required) — asset path relative to the project root.
 * - `width: Option[Float]`
 * - `height: Option[Float]`
 * - `resizeMode: Option[String]` — `"fill"` (default), `"fit"`, `"stretch"`.
 */
public class ImageAdapter private constructor() : FluxAdapter<FluxNativeView> {
    override val kind: String = KIND

    override fun create(nodeId: UInt): FluxNativeView = FluxNativeViewImpl(nodeId, kind)

    override fun update(
        view: FluxNativeView,
        props: Props,
    ) {
        val src = props.getString(PropsIndex.IMAGE_SOURCE)
        if (src.isNullOrEmpty()) {
            // A missing/empty `source` is a load failure: clear the source so the
            // host shows its placeholder rather than a stale bitmap (BR-003).
            if (view.getProperty(PROP_SRC) != null) view.setProperty(PROP_SRC, null)
            if (view.getProperty(PROP_HAS_SRC) != false) view.setProperty(PROP_HAS_SRC, false)
            return
        }
        if (view.getProperty(PROP_SRC) != src) view.setProperty(PROP_SRC, src)
        if (view.getProperty(PROP_HAS_SRC) != true) view.setProperty(PROP_HAS_SRC, true)

        val width = props.getFloat(PropsIndex.IMAGE_WIDTH)
        if (view.getProperty(PROP_WIDTH) != width) view.setProperty(PROP_WIDTH, width)

        val height = props.getFloat(PropsIndex.IMAGE_HEIGHT)
        if (view.getProperty(PROP_HEIGHT) != height) view.setProperty(PROP_HEIGHT, height)

        val mode = props.getString(PropsIndex.IMAGE_RESIZE_MODE) ?: DEFAULT_RESIZE_MODE
        if (view.getProperty(PROP_RESIZE_MODE) != mode) view.setProperty(PROP_RESIZE_MODE, mode)
    }

    override fun setChildren(
        view: FluxNativeView,
        childIds: KList<UInt>,
        children: KList<FluxNativeView>,
    ) {
        // Image is a leaf; the runtime never sends children.
    }

    override fun bindHandler(
        view: FluxNativeView,
        props: Props,
        executor: WeakReference<FluxExecutor>,
    ) {
        // Image has no handlers.
    }

    override fun destroy(view: FluxNativeView) {
        view.setProperty(PROP_SRC, null)
        view.setProperty(PROP_HAS_SRC, false)
        view.clearChildren()
    }

    internal companion object {
        /** The kind tag this adapter handles. Exposed for the factory map. */
        const val KIND: String = "image"

        /** Builds a fresh [ImageAdapter] for one IR node (FLUX-007). */
        fun create(): FluxAdapter<FluxNativeView> = ImageAdapter()

        const val PROP_SRC = "imageSource"
        const val PROP_HAS_SRC = "hasImageSource"
        const val PROP_WIDTH = "imageWidth"
        const val PROP_HEIGHT = "imageHeight"
        const val PROP_RESIZE_MODE = "imageResizeMode"
        const val DEFAULT_RESIZE_MODE = "fill"
    }
}

File: adapters/ui-swift/Sources/FluxUIKit/ImageAdapter.swift
//  ImageAdapter.swift
//  FluxUIKit — `Image` → `UIImageView` (Appendix F.8).

import UIKit

/// Declarative adapter mapping a Flux `Image` node to a `UIImageView`
/// (unified tier; AGENTS.md §3.5).
///
/// Props are read by name; the index is the FNV-1a-32 digest of the name
/// masked to `u16` (`Props.propIndex`), derived identically on server and
/// client (AGENTS.md §3.2) — never a hardcoded positional index. Fields:
/// - `source: String` (required) — asset path relative to the project root,
///   e.g. `"assets/logo.png"`.
/// - `width: Option[Float]`
/// - `height: Option[Float]`
/// - `resizeMode: Option[String]` — `"fill"` (default), `"fit"`, `"stretch"`.
///
/// In dev the bitmap is fetched over HTTP from the dev server's asset route
/// (`http://localhost:7332/assets/<src>`). Load failures (missing asset,
/// offline server, decode error) degrade to a `photo` system placeholder
/// rather than crashing the host — see BR-003. The image view is configured on
/// the main actor because UIKit views may only be touched from the main
/// thread; the network callback re-dispatches to the main actor before
/// mutating the view.
@MainActor
public final class ImageAdapter: FluxAdapter {
    public typealias View = UIImageView
    weak var executor: (any FluxExecutor)?

    /// The dev-server asset base URL (FLUX-019). `<src>` is appended verbatim,
    /// so a node with `src = "assets/logo.png"` resolves to
    /// `…/assets/assets/logo.png`, which the server joins onto the project
    /// root.
    static let assetBaseURL = URL(string: "http://localhost:7332/assets/")!

    /// Shared host-side image cache (FLUX-039): `URLCache` (disk + memory) plus
    /// single-flight fetch. One instance per adapter is unnecessary — a single
    /// app-wide cache keeps every `Image` node from re-fetching the same asset.
    static let cache = ImageCache.shared

    /// Placeholder shown until the bitmap arrives or when loading fails. A
    /// system symbol is used so it is always available and never `nil`.
    private static let placeholder: UIImage = {
        UIImage(systemName: "photo") ?? UIImage()
    }()

    public init(executor: (any FluxExecutor)? = nil) { self.executor = executor }

    public func create() -> UIImageView {
        let imageView = UIImageView()
        imageView.contentMode = .scaleAspectFill
        imageView.clipsToBounds = true
        imageView.image = Self.placeholder
        return imageView
    }

    public func update(_ view: UIImageView, from old: Props, to new: Props) {
        if let width = new.getFloat(named: "width"), let height = new.getFloat(named: "height") {
            view.frame.size = CGSize(width: CGFloat(width), height: CGFloat(height))
        }
        if let mode = new.getString(named: "resizeMode") {
            view.contentMode = Self.contentMode(for: mode)
        }
        guard let src = new.getString(named: "source"), !src.isEmpty else {
            // Missing/empty `source` is treated as a load failure up front: show
            // the placeholder. This is the graceful-degrade path for BR-003.
            view.image = Self.placeholder
            return
        }
        Task { @MainActor in
            await self.load(src, onto: view)
        }
    }

    public func setChildren(_ children: [AnyObject], on view: UIImageView) {
        // `Image` is a leaf; the runtime never sends children.
    }

    public func bindHandler(_ handlerId: FluxHandlerId, to view: UIImageView, nodeId: FluxNodeId) {
        // `Image` has no handlers.
    }

    public func destroy(_ view: UIImageView) {
        view.image = Self.placeholder
    }

    /// Fetches `src` through the shared [ImageCache] (disk + memory, single-flight)
    /// and swaps the decoded bitmap onto `view`, falling back to the placeholder on
    /// any failure. The cache coalesces concurrent same-URL loads and serves
    /// repeats from `URLCache` without a network round-trip. The cache actor is
    /// `await`ed off the main actor; the completion re-dispatches to the main
    /// actor before touching the view (UIKit requirement).
    private func load(_ src: String, onto view: UIImageView) async {
        let url: URL
        do {
            url = try await Self.cache.resolveURL(src, assetBase: Self.assetBaseURL.absoluteString)
        } catch {
            view.image = Self.placeholder
            return
        }
        let result = await Self.cache.get(url)
        // Re-check in-flight identity is unnecessary: the cache is authoritative
        // and a recycled view simply shows the latest resolved image.
        switch result {
        case .success(let data):
            guard let image = UIImage(data: data) else {
                view.image = Self.placeholder
                return
            }
            view.image = image
        case .failure:
            view.image = Self.placeholder
        }
    }

    /// Maps the `contentMode` prop string to a `UIView.ContentMode`.
    private static func contentMode(for mode: String) -> UIView.ContentMode {
        switch mode {
        case "fit": .scaleAspectFit
        case "stretch": .scaleToFill
        default: .scaleAspectFill
        }
    }
}
